src/opaque_closure.c

```diff
@@ -31,17 +17,35 @@ jl_opaque_closure_t *jl_new_opaque_closure(jl_tupletype_t *argt, jl_value_t *isv
     jl_value_t *oc_type JL_ALWAYS_LEAFTYPE;
     oc_type = jl_apply_type2((jl_value_t*)jl_opaque_closure_type, (jl_value_t*)argt, rt_ub);
     JL_GC_PROMISE_ROOTED(oc_type);
-    jl_value_t *captures = NULL;
-    JL_GC_PUSH1(&captures);
+    jl_value_t *captures = NULL, *sigtype = NULL;
+    jl_svec_t *sig_args = NULL;
+    JL_GC_PUSH3(&captures, &sigtype, &sig_args);
     captures = jl_f_tuple(NULL, env, nenv);
+
+    size_t nsig = 1 + jl_svec_len(argt->parameters);
+    sig_args = jl_alloc_svec_uninit(nsig);
+    jl_svecset(sig_args, 0, jl_typeof(captures));
+    for (size_t i = 0; i < nsig-1; ++i) {
+        jl_svecset(sig_args, 1+i, jl_tparam(argt, i));
+    }
+    sigtype = (jl_value_t*)jl_apply_tuple_type_v(jl_svec_data(sig_args), nsig);
+    jl_method_instance_t *mi = jl_specializations_get_linfo((jl_method_t*)source, sigtype, jl_emptysvec);
```

Comment:
@keno may need to review and comment. I thought this function returns the same method-instance each time for a given sigtype, but that the ability to optimize an OC is depending on each copy of the OC potentially having different code depending on where it was initially allocated.

Reply:
Right, we don't currently have the code to that in the optimizer thought. I figured we would do that by allocating a non-cached MethodInstance in the optimizer and putting that in there with the appropriate typed IR. I think that would make this OK, because this specialization cache is then the version that doesn't have any capture optimization done to it.